solr/core/src/java/org/apache/solr/search/MemAllowedLimit.java

```diff
@@ -0,0 +1,130 @@
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadMXBean;
+import java.lang.reflect.Method;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.solr.common.params.CommonParams;
+import org.apache.solr.request.SolrQueryRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MemAllowedLimit implements QueryLimit {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private static final double MEBI = 1024.0 * 1024.0;
+  private static final ThreadMXBean threadBean = ManagementFactory.getThreadMXBean();
+  private static Method GET_BYTES_METHOD;
+  private static boolean supported;
+
+  static {
+    boolean testSupported;
+    try {
+      Class<?> sunThreadBeanClz = Class.forName("com.sun.management.ThreadMXBean");
+      if (sunThreadBeanClz.isAssignableFrom(threadBean.getClass())) {
+        Method m = sunThreadBeanClz.getMethod("isThreadAllocatedMemorySupported");
+        Boolean supported = (Boolean) m.invoke(threadBean);
+        if (supported) {
+          m = sunThreadBeanClz.getMethod("setThreadAllocatedMemoryEnabled", boolean.class);
+          m.invoke(threadBean, Boolean.TRUE);
+          testSupported = true;
+          GET_BYTES_METHOD = sunThreadBeanClz.getMethod("getCurrentThreadAllocatedBytes");
+        } else {
+          testSupported = false;
+        }
+      } else {
+        testSupported = false;
+      }
+    } catch (Exception e) {
+      testSupported = false;
+    }
+    supported = testSupported;
+  }
+
+  private static final ThreadLocal<AtomicLong> threadLocalMem =
+      ThreadLocal.withInitial(() -> new AtomicLong(-1L));
+
+  private long limitBytes;
+  private AtomicLong accumulatedMem = new AtomicLong();
+  private long exitedAt = 0;
+
+  public MemAllowedLimit(SolrQueryRequest req) {
+    if (!supported) {
+      throw new IllegalArgumentException(
+          "Per-thread memory allocation monitoring not available in this JVM.");
+    }
+    float reqMemLimit = req.getParams().getFloat(CommonParams.MEM_ALLOWED, -1.0f);
+    if (reqMemLimit <= 0.0f) {
+      throw new IllegalArgumentException(
+          "Check for limit with hasMemLimit(req) before creating a MemAllowedLimit!");
+    }
+    limitBytes = Math.round(reqMemLimit * MEBI);
+  }
+
+  @VisibleForTesting
+  MemAllowedLimit(float memLimit) {
+    limitBytes = Math.round(memLimit * MEBI);
+  }
+
+  @VisibleForTesting
+  static boolean isSupported() {
+    return supported;
+  }
+
+  static boolean hasMemLimit(SolrQueryRequest req) {
+    return req.getParams().getFloat(CommonParams.MEM_ALLOWED, -1.0f) > 0.0f;
+  }
+
+  @Override
+  public boolean shouldExit() {
+    if (exitedAt > 0L) {
+      return true;
+    }
+
+    try {
+      long currentAllocatedBytes = (Long) GET_BYTES_METHOD.invoke(threadBean);
+      AtomicLong threadMem = threadLocalMem.get();
+      threadMem.compareAndExchange(-1L, currentAllocatedBytes);
```

Comment:
compareAndSet is a more common method with an intuitive name that the reader may not have to look up. Since no return is checked here it would be sufficient.